docker/main/ngen/py_funcs.py

```diff
@@ -0,0 +1,310 @@
+
+    subparsers = parser.add_subparsers(dest='command', help='Specify the Python helper function command to run.')
+    subparsers.required = True
+
+    _parse_for_tar_and_copy(parent_subparsers_container=subparsers)
+    _parse_for_gather_output(parent_subparsers_container=subparsers)
+    _parse_for_move_job_output(parent_subparsers_container=subparsers)
+
+    return parser.parse_args()
+
+
+def _move_to_directory(source_dir: Path, dest_dir: Path, archive_name: Optional[str] = None):
+    """
+    Move source data files from their initial directory to a different directory, potentially combining into an archive.
+
+    Parameters
+    ----------
+    source_dir
+        The source directory containing output files to be placed within the dataset; note that this has already been
+        tested for existence as a directory.
+    dest_dir
+        The destination directory, to which the output data should be moved; note that this path has not yet been tested
+        for existence as a directory.
+    archive_name
+        When files should be archived as part of moving, the name of the archive; when files should not be archived,
+        ``None``.
+    """
+    if not dest_dir.is_dir():
+        raise ValueError(f"{get_date_str()} Can't move job output to non-directory path {dest_dir!s}!")
+
+    if archive_name:
+        logging.info("Archiving output files to output dataset")
+        tar_and_copy(source=source_dir, dest=dest_dir, archive_name=archive_name)
+    else:
+        logging.info("Moving output file(s) to output dataset")
+        for p in source_dir.glob("*"):
+            shutil.move(p, dest_dir)
+
+
+def gather_output(mpi_host_names: List[str], output_write_dir: Path):
+    """
+    Using subprocesses, gather output from remote MPI hosts and collect in the analogous directory on this host.
+
+    Parameters
+    ----------
+    mpi_host_names
+        List of MPI host names for the relevant job.
+    output_write_dir
+        Common job output directory across all hosts, from which data on remotes should be taken and in which data on
+        this host should be collected.
+    """
+    from socket import gethostname
+    local_hostname = gethostname()
+
+    scp_subs = OrderedDict()
+
+    for host in (h for h in mpi_host_names if h != local_hostname):
+        scp_subs[host] = Popen(f"scp -q -r {host}:${output_write_dir!s}/ ${output_write_dir!s}/.")
+
+    for host in scp_subs:
+        return_code = scp_subs[host].wait()
+        if return_code != 0:
+            raise RuntimeError(f"{get_date_str()} gather_output failed for {host} w/ return code {return_code!s}!")
+
+
+def get_date_str() -> str:
+    """
+    Get the current date and time as a string with format ``%Y-%m-%d,%H:%M:%S``
+
+    Returns
+    -------
+    The current date and time as a string.
+    """
+    return datetime.now().strftime('%Y-%m-%d,%H:%M:%S')
+
+
+def move_job_output(output_directory: Path, move_action: str, do_archiving: Optional[bool] = None,
```

Comment:
`Optional[T] = None` is redundant. `do_archiving: Optional[bool] = None` is the same as `do_archiving: bool = None` is the 95% the same thing (that other 5% being when no default value is given - then it's just `Union[bool, None]' and requires the value).

Dropping the `Optional` annotation results in clearer code

Reply:
I disagree. Without `Option[T]` a linter will yell at you for checking if it `is None`. I think the semantics of using an optional boolean as a ternary operation _is_ confusing though. Or at least, im going to look at it twice to realize what's going on. Probably clearest to make this into an `enum` and set the default keyword argument to an enum member.

Reply:
Related to an earlier comment, I think the `Optional[bool]` ternary is preferable to a `force_this: bool, disallow_this: bool`  alternative due to the imprecision and potential contradiction of the latter.  The enum approach seems alright, except that it's causing the scope to keep growing on something intended to be a stopgap because it was already out of scope.  That said, if both of you think the ternary approach is too undesirable (even in the short term), let me know and I can change things to be enum based.

Reply:
FYI, I went ahead and put in the enum solution here.